src/material/schematics/ng-generate/m3-theme/index.ts

```diff
@@ -184,140 +327,86 @@ function createThemeFile(scss: string, tree: Tree, directory?: string) {
 
 export default function (options: Schema): Rule {
   return async (tree: Tree, context: SchematicContext) => {
-    const colorPalettes = getColorTonalPalettes(options.primaryColor);
     let colorComment = 'Color palettes are generated from primary: ' + options.primaryColor;
 
+    // Create tonal palettes for each color and custom color overrides if applicable. Used for both
+    // standard contrast and high contrast schemes since they share the same tonal palettes.
+    const primaryColorHct = getHctFromHex(options.primaryColor);
+    const primaryPalette = TonalPalette.fromHct(primaryColorHct);
+
+    let secondaryPalette;
     if (options.secondaryColor) {
-      colorPalettes.set('secondary', getColorTonalPalettes(options.secondaryColor).get('primary')!);
       colorComment += ', secondary: ' + options.secondaryColor;
+      secondaryPalette = TonalPalette.fromHct(getHctFromHex(options.secondaryColor));
+    } else {
+      secondaryPalette = TonalPalette.fromHueAndChroma(
+        primaryColorHct.hue,
+        Math.max(primaryColorHct.chroma - 32.0, primaryColorHct.chroma * 0.5),
```

Comment:
Where are we getting this math from?

Reply:
I added a comment for it. Its the math Material is doing but we can't create the scheme directly with that class since we allow different palettes to be custom colors and those palettes (except error) are readonly